### ISSUE TITLE
Integration with 🤗 Accelerate or pytorch lightning ignite or catalyst?

### ISSUE THREAD

**Issue by shenmishajing** [open-mmlab/mmengine#619]
First to say, I'm glad to see #583 which says mmengine is considering integration with 🤗 Accelerate library. There is a problem that it's hard to run mm repos like mmdet with multi gpus, multi nodes, or tpus etc. I think there are many researchers like me, are looking forward to something like 🤗 Accelerate library.

But, I have to say, there are too many libraries like this, although most of them work with a trainer unlike 🤗 Accelerate library. I use pytorch lightning with mmdet before the birth of mmengine, so I take pytorch lightning for example. pytorch lightning use a Trainer to run the code, which may cause more codes to change than 🤗 Accelerate library if we want intergration with them. But, it also gives more back.

For example, pytorch lightning support GPU TPU HPU IPU MPS as device, single machine multi machine prem cluster SLURM cluster etc., and half percision as 🤗 Accelerate library. In addition, accelerate is only a part of pytorch lightning, it also support early stop, mentioned by #356 and #597 , profiler, model puring and quantization and many other features. And, yes, it will require so much work to do.

I do not mean that which library is better than others. I open this issue just because of that I think there should not be only voice from one side. Integration with such an accelerate library will take a lot of work, and will be hard to roll back or change to another library if we start it. Therefore, choose one or choose not carefully before we start.

In the end, I'm so glad to see that, whichever library will be chosen or no library will be integrated. The birth of mmengine and the integration with accelerate library will make it the thing I'm looking forward so many days.

**Comment by zhouzaida** [comment]
Hi @shenmishajing , thank you very much for starting this discussion. I'll respond to a few of the questions you've mentioned above.
First of all, it's easy to run multi gpus, multi-machine on mmdet or other OpenMMLab projects, which can be viewed [here](https://mmdetection.readthedocs.io/en/dev-3.x/user_guides/train.html#training-on-multiple-gpus).
Secondly, supporting multiple chips has been in our planning, besides supporting GPU, we also support MPS and are supporting other chips such as NPU, etc. TPU is also in our consideration, but it has been put on hold because we don't have the device at the moment.
Finally, we will work with everyone to make mmengine a full-featured and easy-to-use library.

**Comment by shenmishajing** [comment]
Thanks for your response.

By "hard to run with multi gpus etc." I mean mm repos can not run under those conditions with only change of a few param of runner. Still take pytorch lightning for example, if I want to run my model on gpu, I only have to set accelerator = "gpu" in Trainer. If I want 4 gpus, I just need the device = 4 in addition. Similarly, set accelerator = "tpu" to run on tpus, and mps for apple m chips. For mutli nodes, I only need to set the num_nodes. Even no modification of code have to make if my cluster is managed by SLURM. The method of running an addition shell script will be annoying and confusing, especially when we are debugging with IDE like PyCharm.

It's excited to hear that we will get multi chips support. But, I'm concerned about the way of implement this. Will we integrate    an accelebrate library like 🤗 Accelerate or pytorch lightning or we will do this just by hand?
